### PR TITLE
Audit

### PR DESCRIPTION
**Pull request by malikhaiderkhan**
### Highlights

**Performance before**
s1
a1
s3

**After indexing performance**
s4
a2
s6

### CODE REVIEW

File: queries.sql [hunk 1]
```diff
@@ -1,60 +1,204 @@
...
-SELECT name FROM animals JOIN owners ON animals.owner_id=owners.id WHERE owner_id=4;
-SELECT animals.name FROM animals JOIN species ON animals.species_id=species.id WHERE species_id=1;
-SELECT name,full_name FROM animals RIGHT JOIN owners ON animals.owner_id=owners.id;
-SELECT species.name,COUNT(*) FROM animals JOIN species ON animals.species_id=species.id GROUP BY species.id;
-SELECT animals.name FROM animals JOIN owners ON animals.owner_id=owners.id JOIN species ON animals.species_id=species.id WHERE owner_id=2 AND species_id=2;
-SELECT name FROM animals JOIN owners ON animals.owner_id=owners.id WHERE owner_id=5 AND escape_attempts=0;
-SELECT full_name,COUNT(owner_id) FROM animals JOIN owners ON animals.owner_id=owners.id GROUP BY full_name ORDER BY COUNT(owner_id) DESC LIMIT 1;
+CREATE INDEX idx_animals_id ON visits (animals_id);
+CREATE INDEX idx_vets_id ON visits (vets_id);
+CREATE INDEX idx_email ON owners (email);
```
Comment (shahadat3669) [inline queries.sql]: 👍 Good progress on adding the indexes! However, there's a small adjustment to make to align with the curriculum guidelines.

- [ ] Currently, your index creation operation is inside the `queries.sql` file. According to the [curriculum-databases-projects-template](https://github.com/microverseinc/curriculum-template-databases#:~:text=This%20repository%20includes,Use%20them%20responsibly!), `CREATE` operations like index creation should be inside the `schema.sql` file. Kindly move these operations to the appropriate `schema.sql` file.

By organizing your SQL scripts this way, you'll ensure that your project structure aligns with the curriculum guidelines. Keep up the great work! 👍📜

File: queries.sql [hunk 2]
```diff
@@ -1,60 +1,204 @@
...
-SELECT animals.name,date_of_birth,escape_attempts,neutered,weight_kg,vets.name AS vets_name,vets.age AS vets_age,vets.date_of_graduation AS vets_graduation,date_of_visit FROM visits JOIN vets ON visits.vets_id=vets.id JOIN animals ON visits.animals_id=animals.id ORDER BY date_of_visit DESC LIMIT 1;
-SELECT COUNT(*) FROM visits JOIN vets ON visits.vets_id=vets.id JOIN animals ON visits.animals_id=animals.id JOIN species ON animals.species_id=species.id LEFT JOIN specializations ON vets.id=specializations.vets_id AND species.id=specializations.species_id WHERE specializations.id IS NULL;
-SELECT species.name, COUNT(species.name) AS speciality FROM vets JOIN visits ON vets.id=visits.vets_id JOIN animals ON visits.animals_id=animals.id JOIN species ON animals.species_id=species.id WHERE vets.name='Maisy Smith' GROUP BY species.name ORDER BY speciality DESC LIMIT 1;
+EXPLAIN ANALYZE SELECT COUNT(*) FROM visits where animal_id = 4;
+EXPLAIN ANALYZE SELECT * FROM visits where vet_id = 2;
+EXPLAIN ANALYZE SELECT * FROM owners where email = 'owner_18327@mail.com';
```
Comment (shahadat3669) [inline queries.sql:200]: 👍 Great effort in optimizing your SQL queries! Optimizing queries can be challenging, but you're on the right track. Let's delve into the details:

- [x] Your first query, concerning `animals_id`, appears to be working well. Both before and after creating the index, it returns the same number of rows (10 rows), which is a positive sign. 👍 

- [ ] However, in the second query, focusing on `vet_id`, there's a slight inconsistency. Before creating the index, it shows **5 rows**, but after indexing, it shows **6 rows**. Additionally, it seems you've applied a query limit after indexing. To gain a clearer understanding, I recommend running both queries in the same database, ensuring they operate on the same dataset, and executing them without any query limits. Please provide screenshots of the results for both scenarios.

| Before indexing | After indexing|
| --- | --- |
|  image| image|

- [ ] Regarding the third query, even though you haven't applied a query limit, there's a discrepancy in the results. Before indexing, it shows 8 rows, but after indexing, it returns 4 rows. To investigate this further, I suggest running both queries in the same database, containing an identical dataset, and sharing screenshots of the results for both queries.

| Before indexing | After indexing|
| --- | --- |
|  image| image |

By performing these tests and sharing the results, we can better understand the impact of indexing on your queries. Keep up the excellent work in optimizing your database operations! 👍🔍📊

**Review by shahadat3669 — CHANGES_REQUESTED** [review]
Hi @sulaimonaa, @malikhaiderkhan 

Great job so far! 👏 You've done well, and you're almost there for the next project ⏭️, but there are a few issues 🐛 that you still need to work on.




## Don't Worry 🙏

It's okay to have bugs in your code. 🆗 Please see this code review as an opportunity to learn practical coding skills and discover new tips that can improve your code quality 💯 and help you **get even better** 💼.

**I will guide you on how to address these areas down below 👇.**

## Highlights
✅ Included a pair of screenshots with the results of the  EXPLAIN ANALYZE.
✅ `INSERT INTO` queries are saved in the data.sql file.
✅ Add the requested extra column.

## Required Changes ♻️
_Check the comments under the review._

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._

Cheers and Happy coding!👏👏👏

## Have any doubt ❓

Feel free to leave any questions or comments in the PR thread if something is not 100% clear. Please, remember to tag me in your question so I can receive the notification.

<table width="100%" align="center">
  <tr>
    <td colspan="4" align="center">
      <strong>Contact me: Shahadat Hossain</strong>
    </td>
  </tr>
  <tr>
    <td align="center">
      <a href="https://github.com/shahadata3669">
        GitHub
      </a>
    </td>
    <td align="center">
      <a href="https://www.linkedin.com/in/shahadat-cseng/">
        LinkedIn
      </a>
    </td>
    <td align="center">
      <a href="https://app.slack.com/client/T47CT8XPG/D05GT4E4WAY/rimeto_profile/U04426SD71V">
        Slack
      </a>
    </td>
    <td align="center">
      <a href="mailto:shahadat3669@gmail.com?subject=Microverse%20Code%20Review">
        Email
      </a>
    </td>
  </tr>
</table>

## ⚠️ WARNING ⚠️

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Review by nicupop729 — COMMENTED** [review]
Hello  @team 👋,

_This is an additional review as a result of the second opinion request._

Kudos for the nice effort in this milestone!

For this project, let's only focus on the fact that for the "EXPLAIN ANALYZE SELECT * FROM visits where vet_id = 2;", the query planner output in the "After indexing performance" screenshot is not using the index (there is no Index scan in the output but you can see Seq scan and Filter - please compare this output with the other 2 outputs from the "After" to spot the difference). 
Let's please remove indexes and add them once again. Afterwards, please take screenshots of the `explain analyze` output one more time.

When that is done, please ask for another review from your dashboard.

Keep rocking 🚀!

**Comment by malikhaiderkhan** [comment]
Now after the request on second opinion the reviewer asked me that index scan is not visible in screenshot. So I remove the index and after adding it again now i got the index scan on query "EXPLAIN ANALYZE SELECT * FROM visits where vet_id = 2;" and also the execution time drops by adding indexing. The updated screenshots are also attached above.

**Review by V-Blaze — APPROVED** [review]
Hi @malikhaiderkhan, @sulaimonaa  

# Approved 🟢 🟢 

Your project is complete!  There is nothing else to say other than... it's time to merge it :shipit: 
Congratulations! 🎉 
giphy

## Highlights

- Used `indexes` to decrease execution time 💯
-  Screenshot shows before and after improvemnts👍
-  Nice communication with reviewer 🟢 👍

## Optional suggestions

_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application._

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Pull request merged**